File: src/RNetPi.Core/RNet/SetVolumePacket.cs
using System;
using System.IO;

namespace RNetPi.Core.RNet;

/// <summary>
/// Packet for setting zone volume
/// </summary>
public class SetVolumePacket : EventPacket
{
    public SetVolumePacket(byte controllerID, byte zoneID, int volume)
    {
        if (volume < 0 || volume > 100)
        {
            throw new ArgumentOutOfRangeException(nameof(volume), "Volume must be between 0 and 100");
        }

        TargetPath = new byte[] { 0x02, 0x00 };
        TargetControllerID = controllerID;
        EventID = 0xDE;
        EventTimestamp = (ushort)(volume / 2); // Translate range 0-100 to 0-50
        EventData = zoneID;
        EventPriority = 1;
    }

    public byte GetControllerID()
    {
        return TargetControllerID;
    }

    public byte GetZoneID()
    {
        return (byte)EventData;
    }

    public int GetVolume()
    {
        return EventTimestamp * 2;
    }
}

File: src_js/rnet/packets/SetVolumePacket.js
const EventPacket = require("./EventPacket");

class SetVolumePacket extends EventPacket {
    constructor(controllerID, zoneID, volume) {
        super();

        if (volume < 0 || volume > 100) {
            throw new Error("Volume out of range (0-100) when constructing SetVolumePacket");
        }

        this.targetPath = [0x02, 0x00];
        this.targetControllerID = controllerID;
        this.eventID = 0xDE;
        this.eventTimestamp = Math.floor(volume / 2);  // Translate our range 0-100 to 0-50
        this.eventData = zoneID;
        this.eventPriority = 1;

    }

    getControllerID() {
        return this.targetControllerID;
    }

    getZoneID() {
        return this.eventData;
    }

    getVolume() {
        return this.eventTimestamp * 2;
    }
}

SetVolumePacket.fromPacket = function(eventpacket) {
    if (eventPacket instanceof EventPacket) {
        const powerPacket = new SetPowerPacket();
        eventPacket.copyToPacket(powerPacket);
        return powerPacket;
    }
    else {
        throw new TypeError("Cannot create SetPowerPacket from anything other than an EventPacket");
    }
}

module.exports = SetVolumePacket;
